protocol/messenger_raw_message_resend.go

```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		err = errors.Wrap(err, "Can't resend message with SendCommunityMessage")
+	}
+	m.upsertRawMessageToWatch(rawMessage)
+	return err
+}
+
+func (m *Messenger) handleSendPrivateMessage(rawMessage *common.RawMessage) error {
+	if len(rawMessage.Recipients) == 0 {
+		m.logger.Error("No recipients to resend message", zap.String("id", rawMessage.ID))
+		m.upsertRawMessageToWatch(rawMessage)
+		return errors.New("No recipients to resend message with SendPrivate")
+	}
+
+	var err error
+	for _, r := range rawMessage.Recipients {
+		_, err = m.sender.SendPrivate(context.TODO(), r, rawMessage)
+		if err != nil {
+			err = errors.Wrap(err, fmt.Sprintf("Can't resend message with SendPrivate to %s", common.PubkeyToHex(r)))
+		}
+	}
+
+	m.upsertRawMessageToWatch(rawMessage)
+	return err
+}
+
+func (m *Messenger) handleOtherResendMethods(rawMessage *common.RawMessage) (bool, error) {
+	chat, ok := m.allChats.Load(rawMessage.LocalChatID)
+	if !ok {
+		m.logger.Error("Can't find chat with id", zap.String("id", rawMessage.LocalChatID))
+		return false, nil // Continue with next message if chat not found
+	}
+
+	if !(chat.Public() || chat.CommunityChat()) {
+		return false, nil // Only resend for public or community chats
+	}
+
+	if ok {
+		err := m.persistence.SaveRawMessage(rawMessage)
+		if err != nil {
+			m.logger.Error("Can't save raw message marked as expired", zap.Error(err))
+			return true, err
+		}
+	}
+	return true, m.reSendRawMessage(context.Background(), rawMessage.ID)
+}
+
+func (m *Messenger) shouldResendMessage(message *common.RawMessage, t common.TimeSource) (bool, error) {
+	if m.featureFlags.ResendRawMessagesDisabled {
+		return false, nil
+	}
+	//exponential backoff depends on how many attempts to send message already made
+	power := math.Pow(2, float64(message.SendCount-1))
+	backoff := uint64(power) * uint64(m.config.messageResendMinDelay.Milliseconds())
+	backoffElapsed := t.GetCurrentTime() > (message.LastSent + backoff)
+	return backoffElapsed, nil
+}
+
+// pull a message from the database and send it again
+func (m *Messenger) reSendRawMessage(ctx context.Context, messageID string) error {
+	message, err := m.persistence.RawMessageByID(messageID)
+	if err != nil {
+		return err
+	}
+
+	chat, ok := m.allChats.Load(message.LocalChatID)
+	if !ok {
+		return errors.New("chat not found")
+	}
+
+	_, err = m.dispatchMessage(ctx, common.RawMessage{
+		LocalChatID: chat.ID,
+		Payload:     message.Payload,
+		PubsubTopic: message.PubsubTopic,
+		MessageType: message.MessageType,
+		Recipients:  message.Recipients,
+		ResendType:  message.ResendType,
+		SendCount:   message.SendCount,
+	})
+	return err
+}
+
+// UpsertRawMessageToWatch insert/update the rawMessage to the database, resend it if necessary.
+// relate watch method: Messenger#watchExpiredMessages
+func (m *Messenger) UpsertRawMessageToWatch(rawMessage *common.RawMessage) (*common.RawMessage, error) {
```

Comment:
Noticed that this function is usually invoked after `SendMessageToControlNode` or `SendPrivate` (at least in messenger_communities.go). Is it possible to move this fn's invocation to the aforementioned functions in order to reduce duplication? Or does it rely on some context-specific code?

Reply:
Ok, agree!